Audit at restore time must not throw and fail the operation

Found this in an activity log: 

```xml
<description>NuGet.Protocol.Core.Types.FatalProtocolException: Unable to load the service index for source https://pkgs.dev.azure.com/<redacted>/_packaging/<redacted>/nuget/v3/index.json. ---> System.Net.Http.HttpRequestException: Response status code does not indicate success: 401 (Unauthorized). at System.Net.Http.HttpResponseMessage.EnsureSuccessStatusCode() at NuGet.Protocol.HttpSource.<>c__DisplayClass15_0`1.<<GetAsync>b__0>d.MoveNext() --- 
End of stack trace from previous location where exception was thrown --- at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw() at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task) at NuGet.Common.ConcurrencyUtilities.<ExecuteWithFileLockedAsync>d__5`1.MoveNext() --- 
End of stack trace from previous location where exception was thrown --- at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw() at NuGet.Common.ConcurrencyUtilities.<ExecuteWithFileLockedAsync>d__5`1.MoveNext() --- 
End of stack trace from previous location where exception was thrown --- at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw() at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task) at NuGet.Protocol.HttpSource.<GetAsync>d__15`1.MoveNext() --- 
End of stack trace from previous location where exception was thrown --- at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw() at NuGet.Protocol.ServiceIndexResourceV3Provider.<GetServiceIndexResourceV3>d__11.MoveNext() --- End of inner exception stack trace --- at NuGet.Protocol.ServiceIndexResourceV3Provider.<GetServiceIndexResourceV3>d__11.MoveNext() --- 
End of stack trace from previous location where exception was thrown --- at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw() at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task) at NuGet.Protocol.ServiceIndexResourceV3Provider.<TryCreate>d__10.MoveNext() --- 
End of stack trace from previous location where exception was thrown --- at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw() at NuGet.Protocol.Core.Types.SourceRepository.<GetResourceAsync>d__16`1.MoveNext() --- 
End of stack trace from previous location where exception was thrown --- at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw() at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task) at NuGet.Protocol.Providers.VulnerabilityInfoResourceV3Provider.<TryCreate>d__1.MoveNext() --- 
End of stack trace from previous location where exception was thrown --- at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw() at NuGet.Protocol.Core.Types.SourceRepository.<GetResourceAsync>d__16`1.MoveNext() --- 
End of stack trace from previous location where exception was thrown --- at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw() at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task) at NuGet.Commands.VulnerabilityInformationProvider.<GetVulnerabilityInfoAsync>d__5.MoveNext() --- 
End of stack trace from previous location where exception was thrown --- at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw() at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task) at NuGet.Commands.VulnerabilityInformationProvider.<GetVulnerabilityInformationAsync>d__4.MoveNext() --- 
End of stack trace from previous location where exception was thrown --- at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw() at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task) at NuGet.Commands.Restore.Utility.AuditUtility.<GetAllVulnerabilityDataAsync>d__83.MoveNext() --- 
End of stack trace from previous location where exception was thrown --- at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw() at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task) at NuGet.Commands.Restore.Utility.AuditUtility.<CheckPackageVulnerabilitiesAsync>d__77.MoveNext() --- 
End of stack trace from previous location where exception was thrown --- at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw() at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task) at NuGet.Commands.RestoreCommand.<PerformAuditAsync>d__69.MoveNext() --- 
End of stack trace from previous location where exception was thrown --- at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw() at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task) at NuGet.Commands.RestoreCommand.<ExecuteAsync>d__68.MoveNext() --- 
End of stack trace from previous location where exception was thrown --- at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw() at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task) at NuGet.Commands.RestoreRunner.<ExecuteAsync>d__7.MoveNext() --- 
End of stack trace from previous location where exception was thrown --- at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw() at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task) at NuGet.Commands.RestoreRunner.<ExecuteAndCommitAsync>d__6.MoveNext() --- 
End of stack trace from previous location where exception was thrown --- at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw() at NuGet.SolutionRestoreManager.RestoreOperationLogger.<RunWithProgressAsync>d__35.MoveNext() --- 
End of stack trace from previous location where exception was thrown --- at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw() at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task) at NuGet.SolutionRestoreManager.SolutionRestoreJob.<RestorePackageSpecProjectsAsync>d__26.MoveNext() --- 
End of stack trace from previous location where exception was thrown --- at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw() at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task) at NuGet.SolutionRestoreManager.SolutionRestoreJob.<RestoreAsync>d__24.MoveNext() --- 
End of stack trace from previous location where exception was thrown --- at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw() at NuGet.SolutionRestoreManager.SolutionRestoreJob.<RestoreAsync>d__24.MoveNext() --- 
End of stack trace from previous location where exception was thrown --- at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw() at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task) at NuGet.SolutionRestoreManager.SolutionRestoreJob.<ExecuteAsync>d__23.MoveNext()</description>
```

If you follow the stack trace, you'll see that the throw is happening in restore command project level and throwing all the way to the solution restore job. 

I think the original idea behind this was a warning NU1900, https://learn.microsoft.com/en-us/nuget/reference/errors-and-warnings/nu1900, but we should consider whether need to differentiate between a source being inaccessible altogether and a source not supporting vulnerabilities. 

IMO, we should not log a warning if we can't access the feed altogether, as that'd be surfaced by other means potentially, but that would break some scenarios, so maybe we log a min message instead in that case?
We should log an NU1900 warning if the source has vulnerabilities and they're broken for any reason.

cc @zivkan thoughts?

--- a/src/NuGet.Core/NuGet.Commands/RestoreCommand/VulnerabilityInformationProvider.cs
+++ b/src/NuGet.Core/NuGet.Commands/RestoreCommand/VulnerabilityInformationProvider.cs
@@ -3,6 +3,7 @@
 
 #nullable enable
 
+using System;
 using System.Threading;
 using System.Threading.Tasks;
 using NuGet.Common;
@@ -35,23 +36,31 @@ namespace NuGet.Commands
 
         private async Task<GetVulnerabilityInfoResult?> GetVulnerabilityInfoAsync()
         {
-            IVulnerabilityInfoResource vulnerabilityInfoResource =
-                await _source.GetResourceAsync<IVulnerabilityInfoResource>(CancellationToken.None);
-            if (vulnerabilityInfoResource is null)
+            try
             {
-                return null;
-            }
+                IVulnerabilityInfoResource vulnerabilityInfoResource =
+                    await _source.GetResourceAsync<IVulnerabilityInfoResource>(CancellationToken.None);
+                if (vulnerabilityInfoResource is null)
+                {
+                    return null;
+                }
 
-            // Don't re-use a SourceCacheContext from whatever triggered this, because installing packages
-            // (and a few other scenarios, look at everything that calls SourceCacheContext.MaxAge's setter)
-            // will ignore the http cache, because it wants to allow newly published packages to be installable.
-            // However, in VS, when a new project template installs packages, we don't want to force a re-download
-            // of the vulnerability data. Firstly, from a customer point of view, it's not necessary. The reason
-            // we bust the cache for package install doesn't apply to vulnerability data. Secondly, it triggers
-            // regressions in VS's performance tests.
-            using SourceCacheContext cacheContext = new();
-            GetVulnerabilityInfoResult result = await vulnerabilityInfoResource.GetVulnerabilityInfoAsync(cacheContext, _logger, CancellationToken.None);
-            return result;
+                // Don't re-use a SourceCacheContext from whatever triggered this, because installing packages
+                // (and a few other scenarios, look at everything that calls SourceCacheContext.MaxAge's setter)
+                // will ignore the http cache, because it wants to allow newly published packages to be installable.
+                // However, in VS, when a new project template installs packages, we don't want to force a re-download
+                // of the vulnerability data. Firstly, from a customer point of view, it's not necessary. The reason
+                // we bust the cache for package install doesn't apply to vulnerability data. Secondly, it triggers
+                // regressions in VS's performance tests.
+                using SourceCacheContext cacheContext = new();
+                GetVulnerabilityInfoResult result = await vulnerabilityInfoResource.GetVulnerabilityInfoAsync(cacheContext, _logger, CancellationToken.None);
+                return result;
+            }
+            catch (Exception e)
+            {
+                GetVulnerabilityInfoResult result = new(knownVulnerabilities: null, exceptions: new(e));
+                return result;
+            }
         }
     }
 }

--- a/src/NuGet.Core/NuGet.PackageManagement/AuditUtility.cs
+++ b/src/NuGet.Core/NuGet.PackageManagement/AuditUtility.cs
@@ -137,13 +137,22 @@ namespace NuGet.PackageManagement
 
             static async Task<GetVulnerabilityInfoResult?> GetVulnerabilityInfoAsync(SourceRepository source, SourceCacheContext cacheContext, ILogger logger)
             {
-                IVulnerabilityInfoResource vulnerabilityInfoResource =
-                    await source.GetResourceAsync<IVulnerabilityInfoResource>(CancellationToken.None);
-                if (vulnerabilityInfoResource is null)
+                try
                 {
-                    return null;
+                    IVulnerabilityInfoResource vulnerabilityInfoResource =
+                        await source.GetResourceAsync<IVulnerabilityInfoResource>(CancellationToken.None);
+                    if (vulnerabilityInfoResource is null)
+                    {
+                        return null;
+                    }
+                    return await vulnerabilityInfoResource.GetVulnerabilityInfoAsync(cacheContext, logger, CancellationToken.None);
                 }
-                return await vulnerabilityInfoResource.GetVulnerabilityInfoAsync(cacheContext, logger, CancellationToken.None);
+                catch (Exception exception)
+                {
+                    AggregateException aggregateException = new(exception);
+                    GetVulnerabilityInfoResult result = new(knownVulnerabilities: null, exceptions: aggregateException);
+                    return result;
+                }
             }
         }
 

--- a/test/NuGet.Core.Tests/NuGet.Commands.Test/RestoreCommandTests/VulnerabilityInformationProviderTests.cs
+++ b/test/NuGet.Core.Tests/NuGet.Commands.Test/RestoreCommandTests/VulnerabilityInformationProviderTests.cs
@@ -6,6 +6,7 @@
 using System;
 using System.Threading;
 using System.Threading.Tasks;
+using FluentAssertions;
 using Moq;
 using NuGet.Common;
 using NuGet.Configuration;
@@ -45,6 +46,29 @@ namespace NuGet.Commands.Test.RestoreCommandTests
 
             // Assert
             vulnerabilityResource.Verify(x => x.GetVulnerabilityInfoAsync(It.IsAny<SourceCacheContext>(), It.IsAny<ILogger>(), It.IsAny<CancellationToken>()), Times.Once);
+        }
+
+        // While IVulnerabilityInfoResource.GetVulnerabilityInfoAsync is designed to never throw, the provider needs to get the resource from the service index
+        // and that can throw. Normally it would throw when resolving the graph, but if all packages are already in the global packages folder, then the
+        // SourceRepository instances haven't yet attempted to load the service index. This test ensures that the provider does not throw if the service index
+        // cannot be retrieved, which might happen for reasons like the DNS lookup failed, the service index itself returns an HTTP 404 or 401, network is down, etc.
+        [Fact]
+        public async Task GetVulnerabilityInformationAsync_SourceWithInvalidHostname_DoesNotThrow()
+        {
+            // Arrange
+            // .test is a reserved TLD, so we know it will never exist.
+            var packageSource = new PackageSource("https://nuget.test/v3/index.json");
+            SourceRepository source = Repository.Factory.GetCoreV3(packageSource);
+            var logger = NullLogger.Instance;
+            var provider = new VulnerabilityInformationProvider(source, logger);
+
+            // Act
+            var result = await provider.GetVulnerabilityInformationAsync(CancellationToken.None);
+
+            // Assert
+            result.Should().NotBeNull();
+            result!.KnownVulnerabilities.Should().BeNull();
+            result.Exceptions.Should().NotBeNull();
         }
     }
 }

--- a/test/NuGet.Core.Tests/NuGet.PackageManagement.Test/AuditUtilityTests.cs
+++ b/test/NuGet.Core.Tests/NuGet.PackageManagement.Test/AuditUtilityTests.cs
@@ -520,6 +520,25 @@ namespace NuGet.PackageManagement.Test
         }
 
         [Fact]
+        public async Task GetAllVulnerabilityDataAsync_SourceWithInvalidHost_ReturnResultWithException()
+        {
+            // Arrange
+            // .test is a reserved TLD, so we know it will never exist
+            var packageSource = new PackageSource("https://nuget.test/v3/index.json");
+            SourceRepository source = Repository.Factory.GetCoreV3(packageSource);
+            List<SourceRepository> sourceRepositories = new List<SourceRepository>() { source };
+            using SourceCacheContext cacheContext = new();
+
+            // Act
+            GetVulnerabilityInfoResult? result = await AuditUtility.GetAllVulnerabilityDataAsync(sourceRepositories, cacheContext, NullLogger.Instance, CancellationToken.None);
+
+            // Assert
+            result.Should().NotBeNull();
+            result!.KnownVulnerabilities.Should().BeNull();
+            result.Exceptions.Should().NotBeNull();
+        }
+
+        [Fact]
         public void CreateWarningsForPackagesWithVulnerabilities_CreatesWarningsForAllVulnerabilities()
         {
             var packageA = new PackageIdentity("A", new NuGetVersion(1, 0, 0));

